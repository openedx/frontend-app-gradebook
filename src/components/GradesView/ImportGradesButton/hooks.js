import { useRef } from 'react';
import { selectors, thunkActions } from 'data/redux/hooks';

export const useImportButtonData = () => {
  const gradeExportUrl = selectors.root.useGradeExportUrl();
  const submitImportGradesButtonData = thunkActions.grades.useSubmitImportGradesButtonData();

  const fileInputRef = useRef();

  const handleClickImportGrades = () => fileInputRef.current?.click();
  const handleFileInputChange = () => {
    if (fileInputRef.current?.files[0]) {
      const clearInput = () => {
        fileInputRef.current.value = null;
      };
      const formData = new FormData();
      formData.append('csv', fileInputRef.current.files[0]);
      submitImportGradesButtonData(formData).then(clearInput);
    }
  };

  return {
    fileInputRef,
    gradeExportUrl,
    handleClickImportGrades,
    handleFileInputChange,
  };
};

export default useImportButtonData;
