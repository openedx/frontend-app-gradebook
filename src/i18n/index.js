import { messages as footerMessages } from '@edx/frontend-component-footer';
import { messages as headerMessages } from '@edx/frontend-component-header';
import { messages as paragonMessages } from '@edx/paragon';

import arMessages from './messages/ar.json';
import deMessages from './messages/de.json';
import es419Messages from './messages/es_419.json';
import faIRMessages from './messages/fa_IR.json';
import frMessages from './messages/fr.json';
import frCAMessages from './messages/fr_CA.json';
import hiMessages from './messages/hi.json';
import itMessages from './messages/it.json';
import ptMessages from './messages/pt.json';
import ruMessages from './messages/ru.json';
import ukMessages from './messages/uk.json';
import zhcnMessages from './messages/zh_CN.json';
// no need to import en messages-- they are in the defaultMessage field

const appMessages = {
  ar: arMessages,
  'es-419': es419Messages,
  'fa-ir': faIRMessages,
  fr: frMessages,
  'zh-cn': zhcnMessages,
  pt: ptMessages,
  it: itMessages,
  de: deMessages,
  hi: hiMessages,
  'fr-ca': frCAMessages,
  ru: ruMessages,
  uk: ukMessages,
};

export default [
  paragonMessages,
  footerMessages,
  headerMessages,
  appMessages,
];
