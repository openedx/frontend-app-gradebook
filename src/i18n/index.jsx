import arMessages from './messages/ar.json';
// no need to import en messages-- they are in the defaultMessage field
import es419Messages from './messages/es_419.json';
import frMessages from './messages/fr.json';
import zhcnMessages from './messages/zh_CN.json';

const messages = {
  ar: arMessages,
  'es-419': es419Messages,
  fr: frMessages,
  'zh-cn': zhcnMessages,
};

export default messages;
