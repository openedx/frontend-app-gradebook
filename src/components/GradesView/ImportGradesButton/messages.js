import { defineMessages } from '@edx/frontend-platform/i18n';

const messages = defineMessages({
  csvUploadLabel: {
    id: 'gradebook.BulkManagementHistoryView.csvUploadLabel',
    defaultMessage: 'Upload Grade CSV',
    description: 'A labeled button to upload a CSV containing course grades.',
  },
  importGradesBtnText: {
    id: 'gradebook.GradesView.importGradesBtnText',
    defaultMessage: 'Import Grades',
    description: 'A labeled button to import grades in the BulkManagement Tab File Upload Form',
  },
});

export default messages;
