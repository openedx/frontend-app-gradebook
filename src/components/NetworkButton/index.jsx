/* eslint-disable react/sort-comp, react/button-has-type */
import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { connect } from 'react-redux';

import { StatefulButton, Icon } from '@edx/paragon';
import { FormattedMessage } from '@edx/frontend-platform/i18n';

import selectors from 'data/selectors';
import { StrictDict } from 'utils';

export const buttonStates = StrictDict({
  pending: 'pending',
  default: 'default',
});

/**
 * <NetworkButton />
 * Simplified Download/Upload button, tied to the app's network status (showSpinner).
 * Provides a button with a default `download` icon along with a label, which should
 * be a transifex-friendly message object.  If the `import` argument is passed, the default
 * icon will be an upload icon.
 * If the app is busy on a network task (showSpinner === true), the button will be disabled
 * and show a spinner/working indicator.
 * The buttons can also be passed an optional className for further css customization.
 * @param {string} className - optional extra css class(es)
 * @param {object} label - transifex-friendly message object
 * @param {func} onClick - button on-click action
 * @param {bool} import - should show import icon instead of download icon in default state.
 */
export class NetworkButton extends React.Component {
  get labels() {
    const label = <FormattedMessage {...this.props.label} />;
    return { default: label, pending: label };
  }

  get icons() {
    const iconClass = 'fa mr-2';
    const defaultIcon = this.props.import ? 'fa-upload' : 'fa-download';
    return {
      pending: (<Icon className={classNames(iconClass, 'fa-spinner fa-spin')} />),
      default: (<Icon className={classNames(iconClass, defaultIcon)} />),
    };
  }

  get buttonState() {
    return this.props.showSpinner ? buttonStates.pending : buttonStates.default;
  }

  render() {
    return (
      <StatefulButton
        labels={this.labels}
        variant="outline-primary"
        disabledStates={[buttonStates.pending]}
        className={classNames('ml-2', this.props.className)}
        icons={this.icons}
        state={this.buttonState}
        onClick={this.props.onClick}
      />
    );
  }
}

NetworkButton.defaultProps = {
  className: '',
  showSpinner: false,
  import: false,
};

NetworkButton.propTypes = {
  className: PropTypes.string,
  label: PropTypes.shape({
    id: PropTypes.string,
    defaultMessage: PropTypes.string,
    description: PropTypes.string,
  }).isRequired,
  onClick: PropTypes.func.isRequired,
  import: PropTypes.bool,
  // redux
  showSpinner: PropTypes.bool,
};

export const mapStateToProps = (state) => ({
  showSpinner: selectors.root.shouldShowSpinner(state),
});
export default connect(mapStateToProps)(NetworkButton);
