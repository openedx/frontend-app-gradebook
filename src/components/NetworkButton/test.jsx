import React from 'react';
import { shallow } from 'enzyme';

import { Icon, StatefulButton } from '@edx/paragon';
import { FormattedMessage } from '@edx/frontend-platform/i18n';

import selectors from 'data/selectors';
import { NetworkButton, mapStateToProps, buttonStates } from '.';

jest.mock('@edx/frontend-platform/i18n', () => ({
  FormattedMessage: () => 'FormattedMessage',
}));
jest.mock('@edx/paragon', () => ({
  Icon: () => 'Icon',
  StatefulButton: () => 'StatefulButton',
}));
jest.mock('data/selectors', () => ({
  __esModule: true,
  default: {
    root: { shouldShowSpinner: (state) => ({ shouldShowSpinner: state }) },
  },
}));

describe('NetworkButton component', () => {
  describe('snapshots', () => {
    let el;
    let btnProps;
    const props = {
      className: 'test-class',
      label: {
        id: 'label-id',
        defaultMessage: 'test button label',
        description: 'test button label description',
        showSpinner: false,
      },
    };
    beforeEach(() => {
      props.onClick = jest.fn();
      el = shallow(<NetworkButton {...props} />);
      btnProps = el.find(StatefulButton).props();
    });
    test('snapshot', () => {
      expect(el).toMatchSnapshot();
    });
    it('sets labels to translated label prop', () => {
      expect(btnProps.labels).toEqual({
        default: (<FormattedMessage {...props.label} />),
        pending: (<FormattedMessage {...props.label} />),
      });
    });
    describe('export icons', () => {
      it('sets icons with spinner pending icon and download default', () => {
        expect(btnProps.icons).toEqual({
          pending: (<Icon className="fa mr-2 fa-spinner fa-spin" />),
          default: (<Icon className="fa mr-2 fa-download" />),
        });
      });
    });
    describe('import icons', () => {
      it('sets icons with spinner pending icon and upload default', () => {
        el.setProps({ import: true });
        expect(el.find(StatefulButton).props().icons).toEqual({
          pending: (<Icon className="fa mr-2 fa-spinner fa-spin" />),
          default: (<Icon className="fa mr-2 fa-upload" />),
        });
      });
    });
    describe('buttonState', () => {
      it('is set to pending state if props.showSpinner', () => {
        expect(btnProps.state).toEqual(buttonStates.default);
      });
      it('is set to pending state if props.showSpinner', () => {
        el.setProps({ showSpinner: true });
        expect(el.find(StatefulButton).props().state).toEqual(buttonStates.pending);
        expect(btnProps.state).toEqual(buttonStates.default);
      });
    });
  });
  describe('mapStateToProps', () => {
    const testState = { a: 'wrinkle', in: 'time' };
    let mapped;
    beforeEach(() => {
      mapped = mapStateToProps(testState);
    });
    test('showSpinner from root shouldShowSpinner selector', () => {
      expect(mapped.showSpinner).toEqual(selectors.root.shouldShowSpinner(testState));
    });
  });
});
