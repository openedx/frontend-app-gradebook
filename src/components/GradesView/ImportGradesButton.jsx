/* eslint-disable react/button-has-type, import/no-named-as-default */
import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import { FormattedMessage } from '@edx/frontend-platform/i18n';

import {
  Form,
  FormControl,
  FormGroup,
} from '@edx/paragon';

import selectors from 'data/selectors';
import thunkActions from 'data/thunkActions';
import NetworkButton from 'components/NetworkButton';
import messages from './ImportGradesButton.messages';

/**
 * <ImportGradesButton />
 * File-type input wrapped with hidden control such that when a valid file is
 * added, it is automattically uploaded.
 */
export class ImportGradesButton extends React.Component {
  constructor(props) {
    super(props);
    this.fileInputRef = React.createRef();
    this.handleClickImportGrades = this.handleClickImportGrades.bind(this);
    this.handleFileInputChange = this.handleFileInputChange.bind(this);
  }

  get fileInput() {
    return this.fileInputRef.current;
  }

  get formData() {
    const data = new FormData();
    data.append('csv', this.fileInput.files[0]);
    return data;
  }

  get hasFile() {
    return this.fileInput && this.fileInput.files[0];
  }

  handleClickImportGrades() {
    if (this.fileInput) { this.fileInput.click(); }
  }

  handleFileInputChange() {
    return this.hasFile && (
      this.props.submitImportGradesButtonData(this.formData).then(
        () => { this.fileInput.value = null; },
      )
    );
  }

  render() {
    const { gradeExportUrl } = this.props;
    return (
      <>
        <Form action={gradeExportUrl} method="post">
          <FormGroup controlId="csv">
            <FormControl
              className="d-none"
              type="file"
              label={<FormattedMessage {...messages.csvUploadLabel} />}
              onChange={this.handleFileInputChange}
              ref={this.fileInputRef}
            />
          </FormGroup>
        </Form>

        <NetworkButton
          className="import-grades-btn"
          label={messages.importGradesBtnText}
          onClick={this.handleClickImportGrades}
          import
        />
      </>
    );
  }
}
ImportGradesButton.propTypes = {
  // redux
  gradeExportUrl: PropTypes.string.isRequired,
  submitImportGradesButtonData: PropTypes.func.isRequired,
};

export const mapStateToProps = (state) => ({
  gradeExportUrl: selectors.root.gradeExportUrl(state),
});

export const mapDispatchToProps = {
  submitImportGradesButtonData: thunkActions.grades.submitImportGradesButtonData,
};

export default connect(mapStateToProps, mapDispatchToProps)(ImportGradesButton);
