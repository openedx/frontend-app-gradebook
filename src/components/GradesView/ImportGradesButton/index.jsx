/* eslint-disable react/button-has-type, import/no-named-as-default */
import React from 'react';

import { useIntl } from '@edx/frontend-platform/i18n';

import { Form } from '@edx/paragon';

import NetworkButton from 'components/NetworkButton';
import messages from './messages';
import useImportGradesButtonData from './hooks';

/**
 * <ImportGradesButton />
 * File-type input wrapped with hidden control such that when a valid file is
 * added, it is automattically uploaded.
 */
export const ImportGradesButton = () => {
  const {
    fileInputRef,
    gradeExportUrl,
    handleClickImportGrades,
    handleFileInputChange,
  } = useImportGradesButtonData();
  const { formatMessage } = useIntl();
  return (
    <>
      <Form action={gradeExportUrl} method="post">
        <Form.Group controlId="csv">
          <Form.Control
            data-testid="file-control"
            className="d-none"
            type="file"
            label={formatMessage(messages.csvUploadLabel)}
            onChange={handleFileInputChange}
            ref={fileInputRef}
          />
        </Form.Group>
      </Form>
      <NetworkButton
        className="import-grades-btn"
        label={messages.importGradesBtnText}
        onClick={handleClickImportGrades}
        import
      />
    </>
  );
};
ImportGradesButton.propTypes = {};

export default ImportGradesButton;
