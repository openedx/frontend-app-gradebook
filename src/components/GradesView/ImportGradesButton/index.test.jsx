import React from 'react';
import { shallow } from '@edx/react-unit-test-utils';

import { useIntl } from '@edx/frontend-platform/i18n';
import { Form } from '@edx/paragon';

import NetworkButton from 'components/NetworkButton';
import useImportGradesButtonData from './hooks';
import ImportGradesButton from '.';

jest.mock('components/NetworkButton', () => 'NetworkButton');
jest.mock('./hooks', () => ({ __esModule: true, default: jest.fn() }));

let el;
let props;
describe('ImportGradesButton component', () => {
  beforeAll(() => {
    props = {
      fileInputRef: { current: null },
      gradeExportUrl: 'test-grade-export-url',
      handleClickImportGrades: jest.fn().mockName('props.handleClickImportGrades'),
      handleFileInputChange: jest.fn().mockName('props.handleFileInputChange'),
    };
    useImportGradesButtonData.mockReturnValue(props);
    el = shallow(<ImportGradesButton />);
  });
  describe('behavior', () => {
    it('initializes hooks', () => {
      expect(useImportGradesButtonData).toHaveBeenCalledWith();
      expect(useIntl).toHaveBeenCalledWith();
    });
  });
  describe('render', () => {
    test('snapshot', () => {
      expect(el.snapshot).toMatchSnapshot();
    });
    test('Form', () => {
      expect(el.instance.findByType(Form)[0].snapshot).toMatchSnapshot();
      expect(el.instance.findByType(Form)[0].props.action).toEqual(props.gradeExportUrl);
      expect(el.instance.findByType(Form.Control)[0].props.onChange).toEqual(props.handleFileInputChange);
    });
    test('import button', () => {
      expect(el.instance.findByType(NetworkButton)[0].props.onClick).toEqual(props.handleClickImportGrades);
    });
  });
});
