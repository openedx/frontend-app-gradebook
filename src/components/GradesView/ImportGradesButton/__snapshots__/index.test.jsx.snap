// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImportGradesButton component render Form 1`] = `
<Form
  action="test-grade-export-url"
  method="post"
>
  <Form.Group
    controlId="csv"
  >
    <Form.Control
      className="d-none"
      data-testid="file-control"
      label="Upload Grade CSV"
      onChange={[MockFunction props.handleFileInputChange]}
      type="file"
    />
  </Form.Group>
</Form>
`;

exports[`ImportGradesButton component render snapshot 1`] = `
<Fragment>
  <Form
    action="test-grade-export-url"
    method="post"
  >
    <Form.Group
      controlId="csv"
    >
      <Form.Control
        className="d-none"
        data-testid="file-control"
        label="Upload Grade CSV"
        onChange={[MockFunction props.handleFileInputChange]}
        type="file"
      />
    </Form.Group>
  </Form>
  <NetworkButton
    className="import-grades-btn"
    import={true}
    label={
      Object {
        "defaultMessage": "Import Grades",
        "description": "A labeled button to import grades in the BulkManagement Tab File Upload Form",
        "id": "gradebook.GradesView.importGradesBtnText",
      }
    }
    onClick={[MockFunction props.handleClickImportGrades]}
  />
</Fragment>
`;
